src/connector.c

```diff
@@ -95,50 +113,97 @@ const tlsuv_connector_t* tlsuv_global_connector() {
     return global_connector;
 }
 
-static void on_close(uv_handle_t *h) {
-    struct conn_req_s *req = h->data;
-    tlsuv__free(req);
+static void free_conn_req(struct conn_req_s *cr) {
+    cr->addr ? uv_freeaddrinfo(cr->addr) : 0;
+    tlsuv__free(cr);
 }
 
-static void on_poll_close(uv_handle_t *h) {
-    struct conn_req_s *cr = container_of(h, struct conn_req_s, poll);
-    tlsuv__free(cr);
+static const char *get_name(const struct sockaddr *addr) {
+    static char name[128];
+    if (addr && uv_ip_name(addr, name, sizeof(name)) == 0) { return name; }
+
+    return "<unknown>";
 }
 
-static void on_poll_connect(uv_poll_t *p, int status, int events) {
-    struct conn_req_s *r = container_of(p, struct conn_req_s, poll);
-    uv_poll_stop(p);
+static void connect_work(uv_work_t *work) {
+    struct conn_req_s *cr = container_of(work, struct conn_req_s, connect);
+    int rc = 0;
+    int err = 0;
+    fd_set fds;
 
-    int e;
-    socklen_t len = sizeof(e);
-    getsockopt(r->sock, SOL_SOCKET, SO_ERROR, &e, &len);
+    while (cr->sock == -1) {
+        uv_os_sock_t max_fd = -1;
+        FD_ZERO(&fds);
 
-#if _WIN32
-    switch (e) {
-            case WSAECONNREFUSED: r->error = UV_ECONNREFUSED; break;
-            case WSAECONNRESET: r->error = UV_ECONNRESET; break;
-            case WSAECONNABORTED: r->error = UV_ECONNABORTED; break;
-            default:
-                r->error = -e;
+        for (int i = 0; i < cr->fds; i++) {
+            if (cr->fd[i] == -1) continue;
+
+            FD_SET(cr->fd[i], &fds);
+            max_fd = MAX(max_fd, cr->fd[i] + 1);
+        }
+        if (max_fd == -1) {
+            break;
+        }
+
+        rc = select(max_fd, NULL, &fds, NULL, &(struct timeval){.tv_usec = 250 * 1000});
+        if (cr->cancel) break;
+
+        if (rc == -1) {
+            err = get_error();
+            break;
         }
-#else
-    r->error = -e;
-#endif
 
-    if (r->cancel) {
-        r->error = UV_ECANCELED;
+        if (rc == 0) {
+            UM_LOG(TRACE, "waiting more");
+            continue;
+        }
+
+        for (int i = 0; i < cr->fds; i++) {
```

Comment:
does it make sense to sort or iterate these randomly before finding an IP to use?

Reply:
it won't make a difference -- it tries to connect to all at the same time first to connect is selected